Fix-up to nondet-static option [TG-4365]

majakusber opened the pull request:
This PR makes `nondet-static` option behave in a bit cleaner way:
- we only nondet-initialize custom static fields (before we would also nondet-initialize our variables such inflight_exception etc),
- we replace the original line of CPROVER_init that set the value of a static variable with a line that nondet-initializes it (before we would add the new line after the original assignment).

majakusber commented:
Test-gen pointer bump: https://github.com/diffblue/test-gen/pull/2112

tautschnig commented on src/util/symbol.h:
```diff
@@ -68,6 +68,10 @@ class symbolt
 
+  // Can a static variable be nondet-initialized? User static variables can be
+  // but proprietary ones no (e.g., inflight_exception, clinit_already_run etc)
+  bool is_static_nondet_init = false;
```
This new field needs to be serialised and de-serialiased when writing/reading goto binaries. Also text output needs to be updated.

peterschrammel replied:
Why not just `is_nondet_init` (or some more grammatically correct name)? (The `static` information is already in `is_static_lifetime` if needed).

tautschnig commented on src/goto-instrument/nondet_static.cpp:
```diff
@@ -31,7 +31,7 @@ void nondet_static(
   Forall_goto_program_instructions(i_it, init)
   {
-    const goto_programt::instructiont &instruction=*i_it;
+    const goto_programt::instructiont instruction=*i_it;
```
Why take a copy?

majakusber replied:
In case the instruction will be overriden, I need some info from the original instruction as it's being overriden (and without making a copy it gets into a loop). I could also make a copy of it only at point when I actually need it.

tautschnig commented on src/goto-instrument/nondet_static.cpp:
```diff
@@ -54,7 +54,7 @@ void nondet_static(
         continue;
 
-      i_it=init.insert_before(++i_it);
+      //i_it=init.insert_before(++i_it);
```
This line should just be removed?

majakusber replied:
Yes, it definitely should have been.

tautschnig reviewed (changes requested):
Please review whether adding a flag (and in particular one that doesn't seem to be widely used) to `symbolt` is the right design choice. If yes, more changes are required as indicated below.

peterschrammel commented on jbmc/src/java_bytecode/java_bytecode_convert_class.cpp:
```diff
@@ -635,6 +635,7 @@ void java_bytecode_convert_classt::convert(
     symbolt new_symbol;
 
     new_symbol.is_static_lifetime=true;
+    new_symbol.is_static_nondet_init=true;
```
The commit message could be more informative (e.g. Add is_static_nondet_init to symbol)

smowton replied:
I wouldn't, this will break GBF compatibility across this change, for example

peterschrammel commented on src/goto-instrument/nondet_static.cpp:
```diff
@@ -46,6 +46,10 @@ void nondet_static(
         continue;
 
+      // static that can be nondet initialized?
+      if(!ns.lookup(sym.get_identifier()).is_static_nondet_init)
```
Could this replace the __CPROVER prefix check above? (These variables should all have static lifetime anyway?)

peterschrammel reviewed (commented):
PR seems incomplete at the moment.

majakusber commented:
From design point of view - we discussed this design with Peter, and it sounded like the cleanest solution. Although CI shows that this breaks how `nondet-static` works for C. So option (1) would be to keep the new field and correctly set it for fields in Java (this seems easy) and in C (this does not seem that easy as symbols are being created at many places and it's not obvious which ones are user/private). On the other hand, it seems that in C `nondet-static` was working correctly before the changes so option (2) would be to only fix how it works for Java by adding additional cases in `nondet-static.cpp`, namely if it's `@inflight_exception` or `..::clinit_already_run` or `java.String.Literal` (all these are anyway used in Java only?) do not nondet-initialize it. The advantage of option (2) is that it won't break anything that is already working.

majakusber commented:
@tautschnig @peterschrammel Here is my branch with option (2) implemented: https://github.com/diffblue/cbmc/compare/develop...svorenova:fixup-nondet-static-2

peterschrammel commented:
svorenova:fixup-nondet-static-2 is not an option.

peterschrammel commented:
Looking at it again, I think using an ID_C comment is more appropriate. This is done similarly in comparable situations such as constant and initialization_required.

smowton reviewed (commented):
I'd say almost certainly rather than make a new symbol flag, either maintain the information you need (presumably a set of symbol IDs to reinit) elsewhere, or perhaps tag the symbol using a comment on its type, as e.g. already done to record a method's defining class, here: https://github.com/diffblue/cbmc/blob/develop/jbmc/src/java_bytecode/java_bytecode_convert_method.cpp#L448)

That strategy isn't clean by any stretch of the imagination, but it does at least preserve the serialized (GBF) format and thus only impacts code that uses the new information.

smowton commented:
(The long-term plan is for symbols to have an additional `attributes` irep for this sort of thing)

Commit:
Add a new comment to mark variables that should not be nondet initialized

This will be used by nondet-static option and will mark internal
variables that are not CPROVER variables such as language specific
internal variables - in Java, for example, @inflight_exception and
clinit_already_run

Commit:
Mark internal Java variables with ID_C_no_nondet_initialization

This sets the flag to be true for java::@inflight_exception and static
initializer internal variables (including clinit_already_run)

Commit:
Mark java.lang.String.Literals with ID_C_constant

This will also make sure that nondet-static option does not nondet-
initialize these

Commit:
Make nondet-static check for ID_C_no_nondet_initialization

Commit:
Make nondet-static replace lines in CPROVER_init

Instead of keeping both the original line that sets the value of a
variable and the new line that nondet initializes it, we only keep the
new one

majakusber commented:
I've re-implemented this using a new comment. Let me know what you think.

smowton commented on src/goto-instrument/nondet_static.cpp:
```diff
@@ -38,23 +38,25 @@ void nondet_static(
+      // should not be nondet-initialized, e.g., proprietary variables?
+      if(
+        ns.lookup(sym.get_identifier())
+          .type.get_bool(ID_C_no_nondet_initialization_allowed))
```
Presumably also check for ID_C_constant?

majakusber replied:
Constants are checked below, using `is_constant_or_has_constant_components`

smowton commented:
Better! I liked your suggestion of replacing clinit methods though rather than returning all this stuff to __CPROVER_initialize, did that plan not go anywhere?

majakusber commented:
@smowton Yes, that change will follow in a separate PR. This PR only cleans up nondet-static a bit but we would still call static initializer. The following PR will make sure that even in the static initializer wrapper we only nondet initialize variables.

smowton commented on src/goto-instrument/nondet_static.cpp:
```diff
@@ -38,23 +38,25 @@ void nondet_static(
         continue;
 
+      // should not be nondet-initialized, e.g., proprietary variables?
+      if(
```
Braces around multi-line if

smowton commented on jbmc/src/java_bytecode/java_bytecode_internal_additions.cpp:
```diff
@@ -24,6 +24,7 @@ void java_internal_additions(symbol_table_baset &dest)
     symbol.base_name="__CPROVER_rounding_mode";
     symbol.name=CPROVER_PREFIX "rounding_mode";
     symbol.type=signed_int_type();
+    symbol.type.set(ID_C_no_nondet_initialization_allowed, true);
```
Maybe `ID_C_is_internal_state`? Contrasting user static variables? No strong opinion though.

smowton reviewed (approved):
@peterschrammel @tautschnig maybe you can think of a better name for "this is internal state, nondet-static shouldn't trash it"?

majakusber commented:
@tautschnig @peterschrammel Ready for re-review, please also see Chris's question about the name of the new comment.

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: ec4d5cdd6a30677068d5de2d9ae59cf6e430720c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80742567
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: f0dc8542fe90e4d3641e1bd94945e727b1396cf7).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80752951

martin-cs commented on src/goto-instrument/nondet_static.cpp:
```diff
@@ -38,10 +38,6 @@ void nondet_static(
       const symbol_exprt &sym=to_symbol_expr(
           to_code_assign(instruction.code).lhs());
 
-      // is it a __CPROVER_* variable?
```
If we are going to switch from marking internal / CPROVER generated variables by a CPROVER_PREFIX and instead use this new comment attribute, please could we do it across the whole code-base as I doubt that nondet-static is the only place that makes this distinction.

Also, "change how we mark internal variables" is a bigger change that should be in a PR marked "Fix-up nondet-static".

tautschnig replied:
While I agree with what Martin said (that we should reconsider various `__CPROVER_` prefixed symbols): 1) for several of them the main purpose of the prefix is disambiguation (avoid conflicts with user-provided symbols; 2) let's not ask for huge extra work going well beyond the purpose of the PR (this is happening in several PRs and often stalls progress) - create an issue so that it isn't forgotten instead.

martin-cs replied:
OK; I'm just want to avoid a proliferation of undocumented, slightly disjoint conventions for what are "internal" variables.

martin-cs commented on src/goto-instrument/nondet_static.cpp:
```diff
@@ -38,23 +38,27 @@ void nondet_static(
       if(!ns.lookup(sym.get_identifier()).is_static_lifetime)
         continue;
 
+      // should not be nondet-initialized, e.g., proprietary variables?
```
Terminological nit-picking but ... what is a proprietary variable in an open source system?

martin-cs reviewed (commented):
I don't want to say #751 BUT ... what is the correct way of marking that a variable is "internal only" and what should that mean for different transformations?

tautschnig commented on src/util/irep_ids.def:
```diff
@@ -665,6 +665,7 @@ IREP_ID_TWO(java_lambda_method_handles, lambda_method_handles)
 IREP_ID_ONE(havoc_object)
 IREP_ID_TWO(overflow_shl, overflow-shl)
 IREP_ID_TWO(C_no_initialization_required, #no_initialization_required)
+IREP_ID_TWO(C_no_nondet_initialization_allowed, #no_nondet_initialization_allowed)
```
A bit of nit-picking: "allowed" - who sets those rules? Do I get jailed if I do a nondet-init? :-)  More seriously: The commit message, which is the only documentation here, provides examples, but doesn't really set out the rules. Effectively the commit message will need to set out the semantics.

peterschrammel replied:
> Do I get jailed if I do a nondet-init? 

You get punished with false alarms.

majakusber replied:
I'll leave out the `allowed` to make it clear - just no nondet-init, period, don't even try :) I will also amend the commit message in lines with my last message - marking internal variables that are not CPROVER variables, language specific internal variables.

tautschnig reviewed (approved):
I like this approach. I think there are various details that can (and likely should) be improved as noted by the reviewers, but they should all be reasonably easy to address.

peterschrammel commented:
> replacing clinit methods though rather than returning all this stuff to __CPROVER_initialize, did that plan not go anywhere?

@smowton, the second part of the static nondet fixes are in https://github.com/diffblue/cbmc/pull/2658.
The third part will then de-duplicate the static initialisations that Java does in two different places.

peterschrammel commented:
> better name for "this is internal state

The naming is more explicit on purpose because there might not only be 'internal' variables that must not be nondet-initialised. But I'm also fine with calling it differently.

(since when does Github post when you press Enter... grrrrr)

majakusber commented:
As Peter mentioned the new comment is supposed to identify variables beyond _CPROVER variables that we do not want to nondet-initialize - for example `@inflight_excpetion` is not a CPROVER variable and it probably shouldn't be (it's only used in Java?). Now that I think about it, it makes sense to leave the check for CPROVER variables in nondet-static and really consider the new comment as a mean to mark things in the grey zone between core CPROVER variables and user defined variables such as language specific internal variables. Would you agree?

majakusber commented:
I now pushed the changes along the lines of my previous comment. These should be reflecting all the comments.

I also changed how the last commit works as I learned about `make_skip(), remove_skip(goto_programt)` yesterday ;)

martin-cs commented:
On Fri, 2018-08-03 at 00:28 -0700, svorenova wrote:
> As Peter mentioned the new comment is supposed to identify variables
> beyond _CPROVER variables that we do not want to nondet-initialize -
> for example `@inflight_excpetion` is not a CPROVER variable and it
> probably shouldn't be (it's only used in Java?).

Rounding mode is already a language specific CPROVER variable.  I don't
know what the inflight exception does but if it's part of the model
that CPROVER constructs of how the program works, then maybe it should
be.

>  Now that I think about it, it makes sense to leave the check for
> CPROVER variables in nondet-static and really consider the new
> comment as a mean to mark things in the grey zone between core
> CPROVER variables and user defined variables such as language
> specific internal variables. Would you agree?

That's part of why I asked the question.  I think the behaviour for
CPROVER variables should stay unchanged.  I can see a case for adding
an extra if statement checking for an ID_C_block_nondet_static if you
want additional options of blocking which global variables get non-
det'd (although, is it better as an argument to nondet static?).  This
would be making clear that this is just a flag for that one piece of
functionality and not about a general treatment of "internal"
variables.

Ofcourse, if you want something that marks a global variable as
"internal" (apart from CPROVER prefixes) and affects multiple
subsystems, then that is a different issue and should be addressed
differently.  But I don't think that is the case here?

martin-cs reviewed (approved):
Thanks for the changes; LGTM now.

majakusber commented:
I had to revert the last commit, `make_skip, remove_skip` was misbehaving. @martin-cs That's right, the comment should only affect `nondet-static`.

majakusber commented:
@martin-cs By language specific variable I meant a variable that only exists for particular languages, not necessarily all that we support, such as inflight_exception which in Java determines whether there is an uncaught exception at any given time. I didn't mean that their value is dependent on the chosen language.

martin-cs commented:
On Fri, 2018-08-03 at 03:48 -0700, svorenova wrote:
> @martin-cs By language specific variable I meant a variable that only
> exists for particular languages, not necessarily all that we support,
> such as inflight_exception which in Java determines whether there is
> an uncaught exception at any given time. I didn't mean that their
> value is dependent on the chosen language.

Yes; language specific variables can be and do have CPROVER prefixes.

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: daff1d1c75c5a4795320fe523e6845571dd77397).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80862172
